{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **I. Klebsiella strains** <br>\n",
    "### The point of this part is to go from the set of Klebsiella genomes to a phylogenetic tree.The work was mainly done through the PanACoTA pipeline (https://github.com/gem-pasteur/PanACoTA). Here are the steps : \n",
    "## 1. Download the Klebsiella genomes \n",
    "## 2. Capsule prediction by Kleborate.\n",
    "####  Only the strains with at least a 'good' level of prediction will be kept for the rest of the process \n",
    "#### Screen the resistance genes in order to obtain a virulence score \n",
    "## 3. Annotate the genomes\n",
    "## 4. Generate a pan-genome\n",
    "## 5. Generate a core-genome\n",
    "## 6. Align the core genome\n",
    "## 7. Infer the phylogenetic tree from the previous alignment\n",
    "    \n",
    "https://github.com/gem-pasteur/PanACoTA\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Download the Klebsiella genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The prepare command in order to download all the Kpn sequences from refseq\n",
    "# Changed the 573 to 570 to get all the genus of klebsiella\n",
    "\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=k_indica\n",
    "#SBATCH --partition=short \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=15\n",
    "#SBATCH --mem=75gb \n",
    "#SBATCH --time=1-00:00:00 \n",
    "#SBATCH --output=k_indica%j.log \n",
    "\n",
    "module restore la_base\n",
    "module load singularity\n",
    "\n",
    "singularity run /home/conchae/prediction_depolymerase_tropism/panacota.img prepare -T 570 -o /home/conchae/prediction_depolymerase_tropism/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2. Capsule prediction with Kleborate "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Kleborate step : \n",
    "# Annotation with Kleborate, we only keep the bacteria for which the K-type is clearly identified \n",
    "\n",
    "\n",
    "#Python version : \n",
    "import os\n",
    "from os import system, listdir, mkdir\n",
    "import random\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"\n",
    "\n",
    "\n",
    "# Kleborate small :\n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and specie != \"k_pneumoniae\":\n",
    "        for fasta in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\"), len(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\")):\n",
    "            if fasta[-4:]==\".fna\":\n",
    "                rep=\"_\".join(fasta.split(\"_\")[0:2])\n",
    "                path_in=f\"{path_klebsiella}/{specie}/Database_init\"\n",
    "                path_out=f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}\"\n",
    "                if os.path.isfile(f\"{path_out}/{rep}_Kaptive_out.txt\") == False :\n",
    "                    system(f\"kleborate --kaptive_k --kaptive_k_outfile {path_out}/{rep}_Kaptive_out.txt -a {path_in}/{fasta}\")\n",
    "\n",
    "# Kleborate big :\n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and specie == \"k_pneumoniae\":\n",
    "        for fasta in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\"), len(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\")):\n",
    "            if fasta[-4:]==\".fna\":\n",
    "                rep=\"_\".join(fasta.split(\"_\")[0:2])\n",
    "                path_in=f\"{path_klebsiella}/{specie}/Database_init\"\n",
    "                path_out=f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}\"\n",
    "                if os.path.isfile(f\"{path_out}/{rep}_Kaptive_out.txt\") == False :\n",
    "                    system(f\"kleborate --kaptive_k --kaptive_k_outfile {path_out}/{rep}_Kaptive_out.txt -a {path_in}/{fasta}\")\n",
    "\n",
    "# SH version : \n",
    "#!/bin/bash\n",
    "#BATCH --job-name=kleborate_big\n",
    "#SBATCH --partition=medium \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=40\n",
    "#SBATCH --mem=75gb \n",
    "#SBATCH --time=7-00:00:00 \n",
    "#SBATCH --output=big_kleborate%j.log \n",
    "\n",
    "module restore la_base\n",
    "module load singularity\n",
    "source /storage/apps/ANACONDA/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate kaptive_env\n",
    "\n",
    "python3 /home/conchae/prediction_depolymerase_tropism/script_files/py_file/kleborate_big.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Follow the advancement of the command "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Follow-up script :\n",
    "import os\n",
    "from os import system, listdir, mkdir\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"\n",
    "\n",
    "\n",
    "with open(f\"{path_klebsiella}/completion_kleborate_step.txt\",\"w\") as outfile:\n",
    "    n_total=0\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\"):\n",
    "            n=len(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"))\n",
    "            n_total=n_total+n\n",
    "    n_done=0\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\"):\n",
    "            for rep in os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"):\n",
    "                path_out=f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}\"\n",
    "                if os.path.isfile(f\"{path_out}/{rep}_Kaptive_out.txt\") == True :\n",
    "                    n_done=n_done+1\n",
    "    ratio=(n_done/n_total)*100\n",
    "    if ratio != 1 :\n",
    "        #outfile.write(\"The operation is not completed\")\n",
    "        pass\n",
    "    print(\"The operation is completed at :\", str(ratio), \"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Writting the file ith the Kleborate results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Kleborate resume file : \n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random                                   \n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"     \n",
    "\n",
    "k_specie={\"k_aerogenes\" : \"Klebsiella aerogenes\",\n",
    "          \"k_africana\" : \"Klebsiella africana\",\n",
    "          \"k_grimatii\" : \"Klebsiella grimontii\",\n",
    "          \"k_huaxiensis\" : \"Klebsiella huaxiensis\",\n",
    "          \"k_indica\" : \"Klebsiella indica\",\n",
    "          \"k_michiganesis\" : \"Klebsiella michiganensis\",\n",
    "          \"k_oxytoca\" : \"Klebsiella oxytoca\",\n",
    "          \"k_pasteurii\" : \"Klebsiella pasteurii\",\n",
    "          \"k_pneumoniae\" : \"Klebsiella pneumoniae\",\n",
    "          \"k_quasipneumoniae\": \"Klebsiella quasipneumoniae\",\n",
    "          \"k_quasivariicola\" : \"Klebsiella quasivariicola\",\n",
    "          \"k_spallanzanii\" : \"Klebsiella spallanzanii\",\n",
    "          \"k_variicola\" : \"Klebsiella variicola\"}\n",
    "\n",
    "with open(f\"{path_klebsiella}/kleborate_results_all.tsv\", \"w\") as outfile :\n",
    "    outfile.write(\"Accession \\t Specie \\t K-Serotype \\t Confidence \\t n Missing genes in K-Locus \\t Genome size \\t L90 \\n\")\n",
    "    n_good=0\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "            info_out=\"\"\n",
    "            for file in os.listdir(f\"{path_klebsiella}/{specie}\") :\n",
    "                if file[0:7]==\"LSTINFO\":\n",
    "                    info_out=open(f\"{path_klebsiella}/{specie}/{file}\").read()\n",
    "            for rep in os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"):\n",
    "                info_rep=\"\"\n",
    "                for index, info in enumerate(info_out.split(\"\\n\")) :\n",
    "                    if info.count(rep)>0:\n",
    "                        info_rep=info\n",
    "                kaptive_out=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}/{rep}_Kaptive_out.txt\").read()\n",
    "                first_line=kaptive_out.split(\"\\n\")[0].split(\"\\t\")\n",
    "                secon_line=kaptive_out.split(\"\\n\")[1].split(\"\\t\")\n",
    "                kaptive_dic=dict(zip(first_line,secon_line))\n",
    "                n_miss=len(kaptive_dic[\"Missing expected genes\"].split(\";\"))-1\n",
    "                #gsize=info_rep.split(\"\\t\")[1]\n",
    "                #num_contigs=info_rep.split(\"\\t\")[2]\n",
    "                #L_90=info_rep.split(\"\\t\")[3]\n",
    "                if kaptive_dic[\"Match confidence\"] != \"None\" and kaptive_dic[\"Match confidence\"] != \"Low\":\n",
    "                    #n_good=n_good+1\n",
    "                    outfile.write(f\"{rep} \\t {k_specie[specie]} \\t {kaptive_dic['Best match locus']} \\t {kaptive_dic['Match confidence']} \\t {str(n_miss)} \\n\")\n",
    "    #print(n_good)\n",
    "                        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Compute the resistance for each strain "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Kleborate step : \n",
    "# Annotation with Kleborate, we only keep the bacteria for which the K-type is clearly identified \n",
    "\n",
    "\n",
    "#Python version : \n",
    "import os\n",
    "from os import system, listdir, mkdir\n",
    "import random\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"\n",
    "good_strain=open(f\"{path_klebsiella}/results_kleborate_count.tsv\").read().split(\"\\n\")\n",
    "\n",
    "\n",
    "# Kleborate small :\n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and specie != \"k_pneumoniae\":\n",
    "        for fasta in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\"), len(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\")):\n",
    "            if fasta[-4:]==\".fna\":\n",
    "                rep=\"_\".join(fasta.split(\"_\")[0:2])\n",
    "                path_in=f\"{path_klebsiella}/{specie}/Database_init\"\n",
    "                path_out=f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}\"\n",
    "                if os.path.isfile(f\"{path_out}/{rep}_Kaptive_out.txt\") == False :\n",
    "                    os.system(f\"kleborate --resistance -a {path_in}/{fasta} -o  {path_out}/{rep}_resistance.txt\")\n",
    "\n",
    "# Kleborate big :\n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and specie == \"k_pneumoniae\":\n",
    "        for fasta in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\"), len(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\")):\n",
    "            if fasta[-4:]==\".fna\":\n",
    "                rep=\"_\".join(fasta.split(\"_\")[0:2])\n",
    "                path_in=f\"{path_klebsiella}/{specie}/Database_init\"\n",
    "                path_out=f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}\"\n",
    "                if os.path.isfile(f\"{path_out}/{rep}_Kaptive_out.txt\") == False :\n",
    "                    os.system(f\"kleborate --resistance -a {path_in}/{fasta} -o  {path_out}/{rep}_resistance.txt\")\n",
    "\n",
    "# SH version : \n",
    "#!/bin/bash\n",
    "#BATCH --job-name=kleborate_big\n",
    "#SBATCH --partition=medium \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=40\n",
    "#SBATCH --mem=75gb \n",
    "#SBATCH --time=7-00:00:00 \n",
    "#SBATCH --output=big_kleborate%j.log \n",
    "\n",
    "module restore la_base\n",
    "module load singularity\n",
    "source /storage/apps/ANACONDA/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate kaptive_env\n",
    "\n",
    "python3 /home/conchae/prediction_depolymerase_tropism/script_files/py_file/kleborate_big.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "with open(f\"{path_klebsiella}/kleborate_results_all.tsv\", \"w\") as outfile :\n",
    "    outfile.write(\"Accession \\t Specie \\t K-Serotype \\t Confidence \\t n Missing genes in K-Locus \\t Genome size \\t L90 \\n\")\n",
    "    n_good=0\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "            info_out=\"\"\n",
    "            for file in os.listdir(f\"{path_klebsiella}/{specie}\") :\n",
    "                if file[0:7]==\"LSTINFO\":\n",
    "                    info_out=open(f\"{path_klebsiella}/{specie}/{file}\").read()\n",
    "            for rep in os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"):\n",
    "                info_rep=\"\"\n",
    "                for index, info in enumerate(info_out.split(\"\\n\")) :\n",
    "                    if info.count(rep)>0:\n",
    "                        info_rep=info\n",
    "                kaptive_out=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}/{rep}_Kaptive_out.txt\").read()\n",
    "                first_line=kaptive_out.split(\"\\n\")[0].split(\"\\t\")\n",
    "                secon_line=kaptive_out.split(\"\\n\")[1].split(\"\\t\")\n",
    "                kaptive_dic=dict(zip(first_line,secon_line))\n",
    "                n_miss=len(kaptive_dic[\"Missing expected genes\"].split(\";\"))-1\n",
    "                #gsize=info_rep.split(\"\\t\")[1]\n",
    "                #num_contigs=info_rep.split(\"\\t\")[2]\n",
    "                #L_90=info_rep.split(\"\\t\")[3]\n",
    "                if kaptive_dic[\"Match confidence\"] != \"None\" and kaptive_dic[\"Match confidence\"] != \"Low\":\n",
    "                    #n_good=n_good+1\n",
    "                    outfile.write(f\"{rep} \\t {k_specie[specie]} \\t {kaptive_dic['Best match locus']} \\t {kaptive_dic['Match confidence']} \\t {str(n_miss)} \\n\")\n",
    "    #print(n_good)\n",
    "                "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3. Annotate the genomes\n",
    "Annotation with Prokka "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Prokka annotation : \n",
    "# Python version \n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"\n",
    "\n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and specie == \"k_pneumoniae\":\n",
    "        for fasta in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\"), len(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\")):\n",
    "            if fasta[-4:]==\".fna\":\n",
    "                rep=\"_\".join(fasta.split(\"_\")[0:2])\n",
    "                path_in=f\"{path_klebsiella}/{specie}/Database_init\"\n",
    "                path_out=f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}\"\n",
    "                if os.path.isdir(f\"{path_out}/prokka_annotation\") == False :\n",
    "                    # prokka\n",
    "                    system(f\"prokka {path_in}/{fasta} --norrna --notrna --outdir {path_out}/prokka_annotation --prefix {rep} --compliant --force --cpus 0 \")\n",
    "                                   \n",
    "# ****************************************************************************************************************************************************************                     \n",
    "#!/bin/bash\n",
    "#BATCH --job-name=prokka_small\n",
    "#SBATCH --partition=medium \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=40\n",
    "#SBATCH --mem=75gb \n",
    "#SBATCH --time=7-00:00:00 \n",
    "#SBATCH --output=small_prokka%j.log \n",
    "\n",
    "source /storage/apps/ANACONDA/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate annotation_prokka\n",
    "export PERL5LIB=/home/conchae/.conda/envs/annotation_prokka/bin/perl                                   \n",
    "                                   \n",
    "python3 /home/conchae/prediction_depolymerase_tropism/script_files/py_file/prokka_annotation_small.py\n",
    "                                   \n",
    "cpan install XML::Simple"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check how well the command is going "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random                                   \n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"                                   \n",
    "with open(f\"{path_klebsiella}/completion_prokka.txt\",\"w\") as outfile:\n",
    "    n_total=0\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True :\n",
    "            n=len(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"))\n",
    "            n_total=n_total+n\n",
    "    n_done=0\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True :\n",
    "            for fasta in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\"), len(os.listdir(f\"{path_klebsiella}/{specie}/Database_init\"))):\n",
    "                if fasta[-4:]==\".fna\":\n",
    "                    rep=\"_\".join(fasta.split(\"_\")[0:2])\n",
    "                    path_out=f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}\"\n",
    "                    if os.path.isdir(f\"{path_out}/prokka_annotation_all\") == True :\n",
    "                        n_done=n_done+1\n",
    "    ratio=(n_done/n_total)*100\n",
    "    if ratio != 1 :\n",
    "        #outfile.write(\"The operation is not completed\")\n",
    "        pass\n",
    "    print(\"The operation is completed at :\", str(ratio), \"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 4. Generate a pan-genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# The panacota pangenome step :\n",
    "\n",
    "# The list info file (-l) option :\n",
    "# -l /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/panacota_pangenome_list.txt\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random                                   \n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"     \n",
    "\n",
    "\n",
    "# Third version \n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random                                   \n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"     \n",
    "\n",
    "\n",
    "with open(f\"{path_klebsiella}/panacota_pangenome_list_v2.1.txt\", \"w\") as outfile :\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie==\"k_pneumoniae\" or specie==\"k_quasipneumoniae\" or specie==\"k_variicola\" or specie==\"k_oxytoca\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "            for rep in os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"):\n",
    "                kaptive_out=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}/{rep}_Kaptive_out.txt\").read()\n",
    "                first_line=kaptive_out.split(\"\\n\")[0].split(\"\\t\")\n",
    "                secon_line=kaptive_out.split(\"\\n\")[1].split(\"\\t\")\n",
    "                kaptive_dic=dict(zip(first_line,secon_line))\n",
    "                if kaptive_dic[\"Match confidence\"] != \"None\" and kaptive_dic[\"Match confidence\"] != \"Low\":\n",
    "                    outfile.write(f\"{rep}\\n\")\n",
    "\n",
    "good_strains=open(f\"{path_klebsiella}/panacota_pangenome_list_v2.txt\").read().split(\"\\n\")\n",
    "len(good_strains)\n",
    "# **************************************************************************************************************************************************                    \n",
    "# The dataset_name (-n) option :\n",
    "# -n Klebsiella_genomes \n",
    "\n",
    "# # Getting all the protein files in the directory \n",
    "# -d /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random                                   \n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"   \n",
    "path_prot_files=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "good_strains=open(f\"{path_klebsiella}/panacota_pangenome/panacota_pangenome_list.txt\").read().split(\"\\n\")\n",
    "\n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "        for strain in os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"):\n",
    "            if strain in good_strains :\n",
    "                with open(f\"{path_prot_files}/{strain}.prt\",\"w\") as outfile :\n",
    "                    faa_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.faa\").read().split(\">\")[1:]\n",
    "                    for index, seq in enumerate(faa_file):\n",
    "                        outfile.write(f\">{strain}_{index} {seq}\\n\")\n",
    "                        \n",
    "                        \n",
    "with open(f\"{path_klebsiella}/pangenome_klebsiella_all.prt\",\"w\") as outfile :\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "            for strain in os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"):\n",
    "                if strain in good_strains :\n",
    "                    faa_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.faa\").read().split(\">\")[1:]\n",
    "                    for index, seq in enumerate(faa_file):\n",
    "                        outfile.write(f\">{strain}_{index} {seq}\\n\")\n",
    "\n",
    "\n",
    "# **************************************************************************************************************************************************                    \n",
    "# The pangenome command :\n",
    "\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=panacota_pangenome_2\n",
    "#SBATCH --partition=medium \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=35\n",
    "#SBATCH --mem=75gb \n",
    "#SBATCH --time=7-00:00:00 \n",
    "#SBATCH --output=panacota_v2%j.log \n",
    "\n",
    "module restore la_base\n",
    "module load singularity\n",
    "\n",
    "singularity run /home/conchae/prediction_depolymerase_tropism/panacota.img pangenome  -l /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/panacota_pangenome_list.txt -n Klebsiella_genomes -d /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files -o /home/conchae/prediction_depolymerase_tropism/panacota_pangenome --threads 0 -v \n",
    "\n",
    "\n",
    "OR \n",
    "\n",
    "singularity run /home/conchae/prediction_depolymerase_tropism/panacota.img pangenome  -l /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/panacota_pangenome_list_v2.txt -n Klebsiella_genomes_v2 -d /home/conchae/prediction_depolymerase_tropism/panacota_pangenome_kp_kqp/protein_files -o /home/conchae/prediction_depolymerase_tropism/panacota_pangenome_kp_kqp --threads 0 -v \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 5. Generate a core-genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The panacota core/pers step :\n",
    "\n",
    "# The core/pers command :\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=core_pers\n",
    "#SBATCH --partition=medium \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=35\n",
    "#SBATCH --mem=75gb \n",
    "#SBATCH --time=7-00:00:00 \n",
    "#SBATCH --output=core_pers%j.log \n",
    "\n",
    "module restore la_base\n",
    "module load singularity\n",
    "\n",
    "singularity run /home/conchae/prediction_depolymerase_tropism/panacota.img corepers -p /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/PanGenome-Klebsiella_genomes.All.prt-clust-0.8-mode1-th80.lst -t 0.99 -o /home/conchae/prediction_depolymerase_tropism/panacota_core\n",
    "\n",
    "# Try to use as criteria 95%\n",
    "singularity run /home/conchae/prediction_depolymerase_tropism/panacota.img corepers -p /home/conchae/prediction_depolymerase_tropism/panacota_pangenome_kp_kqp/PanGenome-Klebsiella_genomes_v2.1.All.prt-clust-0.8-mode1-th80.lst -t 0.99 -o /home/conchae/prediction_depolymerase_tropism/panacota_core_kp_kqp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*** \n",
    "### 6. Align the core genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The pre-align step "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The panacota align step :\n",
    "\n",
    "# The core genome file generated in the previous step \n",
    "# -c /home/conchae/prediction_depolymerase_tropism/panacota_core/PersGenome_PanGenome-Klebsiella_genomes.All.prt-clust-0.8-mode1-th80.lst-all_0.99.lst\n",
    "\n",
    "# The list of genomes in the dataset :\n",
    "# -l /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/panacota_pangenome_list.txt\n",
    "\n",
    "# Name of the datset : \n",
    "# -n Klebsiella_genomes \n",
    "\n",
    "# path to the folder containing the directories : \"Proteins\" and \"Genes\"\n",
    "# -d /home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\n",
    "\n",
    "\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random   \n",
    "from Bio.Seq import Seq\n",
    "\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"   \n",
    "path_prot_files=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "path_data=\"/home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\"\n",
    "path_prot=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "good_strains=open(f\"{path_klebsiella}/panacota_pangenome/panacota_pangenome_list.txt\").read().split(\"\\n\")\n",
    "\n",
    "def seq_fasta_format(seq):\n",
    "    seq_1=seq[:]\n",
    "    if seq_1.count(\"\\n\")>0:\n",
    "        seq_1=\"\".join(seq_file.split(\"\\n\")[1:])\n",
    "    else : \n",
    "        (\"The sequence does have any backslashes. Let's continue...\")\n",
    "    seq2=\"\\n\".join(seq_1[i:i+61] for i in range(0,len(seq),61))\n",
    "    return seq2 \n",
    "\n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "        for strain in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"), len(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"))):\n",
    "            if strain in good_strains :\n",
    "                if os.path.isfile(f\"{path_data}/Genes/{strain}.gen\")==False:\n",
    "                    with open(f\"{path_data}/Genes/{strain}.gen\",\"w\") as outfile_ffn :\n",
    "                        used_ffn=[]\n",
    "                        faa_file=open(f\"{path_prot}/{strain}.prt\").read().split(\">\")[1:]\n",
    "                        ffn_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.ffn\").read().split(\">\")[1:]\n",
    "                        for index_prot, seq_prot in enumerate(faa_file):\n",
    "                            header=seq_prot.split(\"\\n\")[0]\n",
    "                            fasta_faa=\"\".join(seq_prot.split(\"\\n\")[1:])\n",
    "                            func_faa=\"\".join(seq_prot.split(\"\\n\")[0].split(\" \")[1:])\n",
    "                            index_max=len(ffn_file)-len(faa_file)+index_prot\n",
    "                            for index_gen in range(index_prot,index_max) :\n",
    "                                fasta_ffn=\"\".join(ffn_file[index_gen].split(\"\\n\")[1:])\n",
    "                                translation=str(Seq(fasta_ffn).translate())[0:-1].strip()\n",
    "                                header_ffn=ffn_file[index_gen].split(\"\\n\")[0]\n",
    "                                if len(fasta_ffn)%3 ==0 and len(fasta_faa)==len(translation):\n",
    "                                    if translation==fasta_faa and header_ffn not in used_ffn:\n",
    "                                        break\n",
    "                            print(\"We might have it now :\\n\",strain,\"\\n\",index_prot,\"\\n\")\n",
    "                            outfile_ffn.write(f\">{header}\\n{seq_fasta_format(fasta_ffn)}\\n\")\n",
    "\n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "        for strain in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"), len(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"))):\n",
    "            if strain in good_strains :\n",
    "                if os.path.isfile(f\"{path_data}/Genes/{strain}.gen\")==False:\n",
    "                    print(\"Something wrong with that one :\",strain)\n",
    "                            \n",
    "# The pre_align commd :\n",
    "\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=pre_align\n",
    "#SBATCH --partition=medium \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=40\n",
    "#SBATCH --mem=75gb \n",
    "#SBATCH --time=3-00:00:00 \n",
    "#SBATCH --output=pre_align%j.log \n",
    "\n",
    "module restore la_base\n",
    "source  /home/conchae/bio_module/bin/activate\n",
    "\n",
    "python3 /home/conchae/prediction_depolymerase_tropism/script_files/py_file/panacota_pre_align.py\n",
    "\n",
    "# Follow the command : \n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random   \n",
    "\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"   \n",
    "path_prot_files=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "path_data=\"/home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\"\n",
    "path_prot=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "good_strains=open(f\"{path_klebsiella}/panacota_pangenome/panacota_pangenome_list.txt\").read().split(\"\\n\")\n",
    "\n",
    "n_done= len(os.listdir(f\"{path_data}/Genes\"))\n",
    "n_total= len(good_strains)\n",
    "\n",
    "print(\"The process has been done at : \",(n_done/n_total)*100,\"%\")\n",
    "    \n",
    "\n",
    "# Output directory :\n",
    "# -o /home/conchae/prediction_depolymerase_tropism/panacota_align\n",
    "\n",
    "# Generating the right files for the align step ...\n",
    "\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"   \n",
    "path_prot_files=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "path_data=\"/home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\"\n",
    "path_prot=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "good_strains=open(f\"{path_klebsiella}/panacota_pangenome/panacota_pangenome_list.txt\").read().split(\"\\n\")\n",
    "\n",
    "path_pana_core=\"/home/conchae/prediction_depolymerase_tropism/panacota_core\"\n",
    "            \n",
    "info_core=open(f\"{path_data}/core_genes.txt\").read().split(\"\\n\")\n",
    "for index_core,line in enumerate(info_core):\n",
    "    strain=line.split(\"\\t\")[0]\n",
    "    strains_target=[target for target in line.split(\"\\t\")[1:] if target]\n",
    "    test_assay=[]\n",
    "    if os.path.isfile(f\"{path_data}/Genes_5/{strain}.gen\")==False:\n",
    "        with open(f\"{path_data}/Genes_5/{strain}.gen\", \"w\") as outfile :\n",
    "            specie=\"\"\n",
    "            for species in os.listdir(path_klebsiella):\n",
    "                if species[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{species}/refseq/bacteria/{strain}\")== True :\n",
    "                    specie=species\n",
    "                    break\n",
    "            faa_file=[faa for faa in open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\") if faa]\n",
    "            ffn_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.ffn\").read().split(\">\")[1:]\n",
    "            for index_faa, line_faa in enumerate(faa_file):\n",
    "                strain_id=line_faa.split(\" \")[0]\n",
    "                prot_tag=line_faa.split(\" \")[1]\n",
    "                if strain_id in strains_target :\n",
    "                    for index_ffn, line_ffn in enumerate(ffn_file) :\n",
    "                        if line_ffn.count(prot_tag)>0 :\n",
    "                            seq_ffn=\"\\n\".join(line_ffn.split(\"\\n\")[1:])\n",
    "                            header=f\">{strain_id} {prot_tag}\\n{seq_ffn}\"\n",
    "                            test_assay.append(prot_tag)\n",
    "                            outfile.write(header)\n",
    "            if len(test_assay)==len(strains_target) :\n",
    "                print(\"This strain is ok\")\n",
    "            else :\n",
    "                print(\"This strain has gotten :\", len(test_assay), \"instead of : \",len(strains_target))\n",
    "\n",
    "# delete the wrong ones :\n",
    "info_core=open(f\"{path_data}/core_genes.txt\").read().split(\"\\n\")\n",
    "for index_core,line in enumerate(info_core):\n",
    "    strain=line.split(\"\\t\")[0]\n",
    "    strains_target=[target for target in line.split(\"\\t\")[1:] if target]\n",
    "    test_assay=[]\n",
    "    if os.path.isfile(f\"{path_data}/Genes_/{strain}.gen\")==True:\n",
    "        file=open(f\"{path_data}/Genes_5/{strain}.gen\").read().split(\">\")[1:]\n",
    "        if len(file) != len(strains_target)/2 :\n",
    "            print(strain, len(file), len(strains_target))\n",
    "            system(f\"rm {path_data}/Genes_5/{strain}.gen\")\n",
    "        else :\n",
    "            print(\"Ok let's go\", strain)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> The align step "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The align command :\n",
    "\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=align\n",
    "#SBATCH --partition=medium \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=40\n",
    "#SBATCH --mem=100gb \n",
    "#SBATCH --time=7-00:00:00 \n",
    "#SBATCH --output=align%j.log \n",
    "\n",
    "module restore la_base\n",
    "module load singularity\n",
    "\n",
    "\n",
    "singularity run /home/conchae/prediction_depolymerase_tropism/panacota.img align  -c /home/conchae/prediction_depolymerase_tropism/panacota_core/PersGenome_PanGenome-Klebsiella_genomes.All.prt-clust-0.8-mode1-th80.lst-all_0.99.lst -l /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/panacota_pangenome_list.txt -n Klebsiella_genomes -d /home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes -o /home/conchae/prediction_depolymerase_tropism/panacota_align --threads 0 -v \n",
    "\n",
    "\n",
    "singularity run /home/conchae/prediction_depolymerase_tropism/panacota.img align \n",
    "-c /home/conchae/prediction_depolymerase_tropism/panacota_core/PersGenome_PanGenome-Klebsiella_genomes.All.prt-clust-0.8-mode1-th80.lst-all_0.99.lst\n",
    "-l /home/conchae/prediction_depolymerase_tropism/panacota_pangenome/panacota_pangenome_list.txt\n",
    "-n Klebsiella_genomes\n",
    "-d /home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\n",
    "-o /home/conchae/prediction_depolymerase_tropism/panacota_align\n",
    "--threads 0\n",
    "-v\n",
    "-P\n",
    "\n",
    "# something wrong : \n",
    "\n",
    "#log file :\n",
    "\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random   \n",
    "from Bio.Seq import Seq\n",
    "\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"   \n",
    "path_data=\"/home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\"\n",
    "path_prot=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "path_align_log=\"/home/conchae/prediction_depolymerase_tropism/panacota_align/PanACoTA-align_Klebsiella_genomes.log.err\"\n",
    "path_align=\"/home/conchae/prediction_depolymerase_tropism/panacota_align\"\n",
    "path_list_fam=\"/home/conchae/prediction_depolymerase_tropism/panacota_core/PersGenome_PanGenome-Klebsiella_genomes.All.prt-clust-0.8-mode1-th80.lst-all_0.99.lst\"\n",
    "\n",
    "good_strains=open(f\"{path_klebsiella}/panacota_pangenome/panacota_pangenome_list.txt\").read().split(\"\\n\")\n",
    "\n",
    "log_file=open(path_align_log).read().split(\"\\n\")[0:-2]\n",
    "print(len(log_file))\n",
    "for index, line in enumerate(log_file):\n",
    "    if line :\n",
    "        start=line.split(\" \")[12][1:-1]\n",
    "        stop=line.split(\" \")[16][1:-1]\n",
    "        n=int(start)+int(stop)\n",
    "        print(n,start,stop)\n",
    "        \n",
    "list_families=open(path_list_fam).read()\n",
    "\n",
    "# Is the len of the .prt the same as the .gen files ?\n",
    "# No : \n",
    "for specie in os.listdir(path_klebsiella):\n",
    "    if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "        for strain in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"), len(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"))):\n",
    "            if strain in good_strains :\n",
    "                if os.path.isfile(f\"{path_data}/Genes/{strain}.gen\")==False and os.path.isfile(f\"{path_data}/Proteins/{strain}.prt\")==False:\n",
    "                    print(\"Something is DEEPLY wrong with this kid : \",specie, strain)\n",
    "                elif os.path.isfile(f\"{path_data}/Genes/{strain}.gen\")==False :\n",
    "                    print(\"Gene file is wrong for him : \",specie, strain)\n",
    "                elif os.path.isfile(f\"{path_data}/Proteins/{strain}.prt\")==False :\n",
    "                    print(\"Prot file is wrong for him : \",specie, strain)\n",
    "                faa_file=open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\")\n",
    "                ffn_file=open(f\"{path_data}/Genes/{strain}.gen\").read().split(\">\")\n",
    "                if len(ffn_file) != len(faa_file) :\n",
    "                    print(\"Wrong with this one : \",len(ffn_file),len(faa_file) ,specie, strain)\n",
    "                else : \n",
    "                    pass\n",
    "\n",
    "wrong_files={}\n",
    "wrong_txt=open(f\"{path_align}/wrong_files\").read().split(\"\\n\")\n",
    "for index, line in enumerate(wrong_txt):\n",
    "    if line :\n",
    "        strain=line.split(\" \")[-1]\n",
    "        specie=line.split(\" \")[-2]\n",
    "        wrong_files[strain]=specie\n",
    "    \n",
    "for strain in wrong_files:\n",
    "    specie=wrong_files[strain]\n",
    "    faa_file=open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\")\n",
    "    ffn_file=open(f\"{path_data}/Genes/{strain}.gen\").read().split(\">\")\n",
    "    if len(ffn_file) != len(faa_file) :\n",
    "        print(\"Following strain is wrong :\",specie, strain, \"\\n\",len(faa_file), len(ffn_file))\n",
    "        with open(f\"{path_data}/Genes/{strain}.gen\",\"w\") as outfile_ffn :\n",
    "            used_ffn=[]\n",
    "            faa_file=open(f\"{path_prot}/{strain}.prt\").read().split(\">\")[1:]\n",
    "            ffn_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.ffn\").read().split(\">\")[1:]\n",
    "            for index_prot, seq_prot in enumerate(faa_file):\n",
    "                header=seq_prot.split(\"\\n\")[0]\n",
    "                fasta_faa=\"\".join(seq_prot.split(\"\\n\")[1:])\n",
    "                func_faa=\"\".join(seq_prot.split(\"\\n\")[0].split(\" \")[1:])\n",
    "                index_max=len(ffn_file)-len(faa_file)+index_prot\n",
    "                for index_gen in range(index_prot,index_max) :\n",
    "                    fasta_ffn=\"\".join(ffn_file[index_gen].split(\"\\n\")[1:])\n",
    "                    translation=str(Seq(fasta_ffn).translate())[0:-1].strip()\n",
    "                    header_ffn=ffn_file[index_gen].split(\"\\n\")[0]\n",
    "                    if len(fasta_ffn)%3 ==0 and len(fasta_faa)==len(translation):\n",
    "                        if translation==fasta_faa and header_ffn not in used_ffn:\n",
    "                            break\n",
    "                print(\"We might have it now :\\n\",specie,strain,\"\\n\",index_prot,\"\\n\")\n",
    "                outfile_ffn.write(f\">{header}\\n{seq_fasta_format(fasta_ffn)}\\n\")\n",
    "    else :\n",
    "        print(\"Seems like \",specie, strain, \"is fixed\")\n",
    "\n",
    "for strain in wrong_files:\n",
    "    specie=wrong_files[strain]\n",
    "    with open(f\"{path_data}/Genes/{strain}.gen\",\"w\") as outfile_ffn :\n",
    "        used_ffn=[]\n",
    "        faa_file=open(f\"{path_prot}/{strain}.prt\").read().split(\">\")[1:]\n",
    "        ffn_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.ffn\").read().split(\">\")[1:]\n",
    "        for index_prot, seq_prot in enumerate(faa_file):\n",
    "            header=seq_prot.split(\"\\n\")[0]\n",
    "            fasta_faa=\"\".join(seq_prot.split(\"\\n\")[1:])\n",
    "            func_faa=\"\".join(seq_prot.split(\"\\n\")[0].split(\" \")[1:])\n",
    "            index_max=len(ffn_file)-len(faa_file)+index_prot\n",
    "            for index_gen in range(index_prot,index_max) :\n",
    "                fasta_ffn=\"\".join(ffn_file[index_gen].split(\"\\n\")[1:])\n",
    "                translation=str(Seq(fasta_ffn).translate())[0:-1].strip()\n",
    "                header_ffn=ffn_file[index_gen].split(\"\\n\")[0]\n",
    "                if len(fasta_ffn)%3 ==0 and len(fasta_faa)==len(translation):\n",
    "                    if translation==fasta_faa and header_ffn not in used_ffn:\n",
    "                        break\n",
    "            print(\"We might have it now :\\n\",specie,strain,\"\\n\",index_prot,\"\\n\")\n",
    "            outfile_ffn.write(f\">{header}\\n{seq_fasta_format(fasta_ffn)}\\n\")\n",
    "# Are all the genomes in the good_strains have their file in the data folder ?\n",
    "# Yes\n",
    "'''for genome in good_strains:\n",
    "    if os.path.isfile(f\"{path_data}/Genes/{genome}.gen\")== False or os.path.isfile(f\"{path_data}/Proteins/{genome}.prt\")== False:\n",
    "        print(genome)'''\n",
    "\n",
    "\n",
    "# The proteins from the same family do not have the same lengths\n",
    "\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random   \n",
    "from Bio.Seq import Seq\n",
    "\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"   \n",
    "path_data=\"/home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\"\n",
    "path_prot=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "path_align_log=\"/home/conchae/prediction_depolymerase_tropism/panacota_align/PanACoTA-align_Klebsiella_genomes.log.err\"\n",
    "path_align=\"/home/conchae/prediction_depolymerase_tropism/panacota_align\"\n",
    "path_list_fam=\"/home/conchae/prediction_depolymerase_tropism/panacota_core/PersGenome_PanGenome-Klebsiella_genomes.All.prt-clust-0.8-mode1-th80.lst-all_0.99.lst\"\n",
    "\n",
    "wrong_genes={}\n",
    "for file in os.listdir(f\"{path_data}/Genes\"):\n",
    "    strain=file.split(\".gen\")[0]\n",
    "    cont_genes=[]\n",
    "    wrong_genes[strain]=cont_genes\n",
    "    faa_file=open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\")[1:]\n",
    "    ffn_file=open(f\"{path_data}/Genes/{strain}.gen\").read().split(\">\")[1:]\n",
    "    for index_gen, seq_gen in enumerate(ffn_file):\n",
    "        header_seq=seq_gen.split(\"\\n\")[0]\n",
    "        for index_prt, seq_prot in enumerate(faa_file):\n",
    "            header_prt=seq_prot.split(\"\\n\")[0]\n",
    "            if header_prt == header_seq :\n",
    "                fasta_ffn=\"\".join(seq_gen.split(\"\\n\")[1:])\n",
    "                fasta_faa=\"\".join(seq_prot.split(\"\\n\")[1:])\n",
    "                translation=str(Seq(fasta_ffn).translate())[0:-1].strip()\n",
    "                if translation == fasta_faa :\n",
    "                    pass\n",
    "                else :\n",
    "                    wrong_genes[strain].append(header_seq)\n",
    "                    print(\"That one is wrong :\", strain , header_seq ,\"\\n\", fasta_faa , \"\\n\",translation)\n",
    "                        \n",
    "                    \n",
    "# *********************************************************************************************\n",
    "# It's all wrong, let's revise that  :\n",
    "\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random   \n",
    "from Bio.Seq import Seq\n",
    "\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"   \n",
    "path_prot_files=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "path_data=\"/home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\"\n",
    "path_prot=\"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/protein_files\"\n",
    "good_strains=open(f\"{path_klebsiella}/panacota_pangenome/panacota_pangenome_list.txt\").read().split(\"\\n\")\n",
    "\n",
    "def seq_fasta_format(seq):\n",
    "    seq_1=seq[:]\n",
    "    if seq_1.count(\"\\n\")>0:\n",
    "        seq_1=\"\".join(seq_file.split(\"\\n\")[1:])\n",
    "    else : \n",
    "        (\"The sequence does have any backslashes. Let's continue...\")\n",
    "    seq2=\"\\n\".join(seq_1[i:i+61] for i in range(0,len(seq),61))\n",
    "    return seq2 \n",
    "\n",
    "for strain in wrong_files:\n",
    "    specie=wrong_files[strain]\n",
    "    faa_file=open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\")\n",
    "    ffn_file=open(f\"{path_data}/Genes/{strain}.gen\").read().split(\">\")\n",
    "    if len(ffn_file) != len(faa_file) :\n",
    "        print(\"Following strain is wrong :\",specie, strain, \"\\n\",len(faa_file), len(ffn_file))\n",
    "        with open(f\"{path_data}/Genes/{strain}.gen\",\"w\") as outfile_ffn :\n",
    "            used_ffn=[]\n",
    "            faa_file=open(f\"{path_prot}/{strain}.prt\").read().split(\">\")[1:]\n",
    "            ffn_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.ffn\").read().split(\">\")[1:]\n",
    "            for index_prot, seq_prot in enumerate(faa_file):                   # Creating the .gen file :\n",
    "                with open(f\"{path_data}/Genes_2/{strain}.gen\",\"w\") as outfile_ffn :\n",
    "                    used_ffn=[]\n",
    "                    faa_file=open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\")[1:]\n",
    "                    ffn_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.ffn\").read().split(\">\")[1:]\n",
    "                    # Going through the protein file\n",
    "                    for index_prot, seq_prot in enumerate(faa_file):\n",
    "                        header=seq_prot.split(\"\\n\")[0]\n",
    "                        fasta_faa=\"\".join(seq_prot.split(\"\\n\")[1:])\n",
    "                        func_faa=\"\".join(seq_prot.split(\"\\n\")[0].split(\" \")[1:])\n",
    "                        index_max=len(ffn_file)-len(faa_file)+index_prot\n",
    "                        index_min=index_prot-10\n",
    "                        for index_gen in range(index_min,index_max) :\n",
    "                            fasta_ffn=\"\".join(ffn_file[index_gen].split(\"\\n\")[1:])\n",
    "                            translation=str(Seq(fasta_ffn).translate())[0:-1].strip()\n",
    "                            header_ffn=ffn_file[index_gen].split(\"\\n\")[0]\n",
    "                            if len(fasta_ffn)%3 ==0 and len(fasta_faa)==len(translation):\n",
    "                                if translation==fasta_faa and header_ffn not in used_ffn:\n",
    "                                    print(fasta_faa, \"\\n\", translation , \"\\n\\n\", fasta_ffn, \"\\n\\n\\n\\n\")\n",
    "                                    used_ffn.append(header_ffn)\n",
    "                                    outfile_ffn.write(f\">{header}\\n{seq_fasta_format(fasta_ffn)}\\n\")\n",
    "                                    break\n",
    "                                else :\n",
    "                                    print(\"Nothing to see here.\")\n",
    "                        print(\"We might have it now :\\n\",strain,\"\\n\",index_prot,\"\\n\")\n",
    "\n",
    "with open(f\"{path_data}/gen_files_control.tsv\",\"a+\") as outfile_c :\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        # Test on a specific specie first\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "            for strain in random.sample(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"), len(os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"))):\n",
    "                if strain in good_strains :\n",
    "                    if os.path.isfile(f\"{path_data}/Genes/{strain}.gen\")==False:\n",
    "                        # Creating the .gen file :\n",
    "                        with open(f\"{path_data}/Genes/{strain}.gen\",\"w\") as outfile_ffn :\n",
    "                            used_ffn=()\n",
    "                            all_headers_faa=()\n",
    "                            used_headers_faa=()\n",
    "                            headers_ffn=()\n",
    "                            faa_file=open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\")[1:]\n",
    "                            ffn_file=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{strain}/prokka_annotation_all/{strain}.ffn\").read().split(\">\")[1:]\n",
    "                            expected_len=len(faa_file)\n",
    "                            # Going through the protein file\n",
    "                            for index_prot, seq_prot in enumerate(faa_file):\n",
    "                                header=seq_prot.split(\"\\n\")[0]\n",
    "                                fasta_faa=\"\".join(seq_prot.split(\"\\n\")[1:]).strip()\n",
    "                                func_faa=\"\".join(seq_prot.split(\"\\n\")[0].split(\" \")[1:])\n",
    "                                index_max=len(faa_file)\n",
    "                                index_min=index_prot-250\n",
    "                                all_headers_faa=all_headers_faa+(header,)\n",
    "                                for index_gen in range(index_min,index_max) :\n",
    "                                    fasta_ffn=\"\".join(ffn_file[index_gen].split(\"\\n\")[1:])\n",
    "                                    translation=str(Seq(fasta_ffn).translate())[0:-1].strip()\n",
    "                                    header_ffn=ffn_file[index_gen].split(\"\\n\")[0]\n",
    "                                    if len(fasta_ffn)%3 ==0 and len(fasta_faa)==len(translation):\n",
    "                                        if translation==fasta_faa and header_ffn not in used_ffn:\n",
    "                                            #print(fasta_faa, \"\\n\", translation , \"\\n\\n\", fasta_ffn, \"\\n\\n\\n\\n\")\n",
    "                                            used_ffn=used_ffn+(header_ffn,)\n",
    "                                            used_headers_faa=used_headers_faa+(header,)\n",
    "                                            #print(\"Looks good here\",\"\\n\",header)\n",
    "                                            outfile_ffn.write(f\">{header}\\n{seq_fasta_format(fasta_ffn)}\\n\")\n",
    "                                            break\n",
    "                                        else :\n",
    "                                            #print(\"Nothing to see here. But maybe ...\",strain)\n",
    "                                            #print(translation,\"\\n\",fasta_faa,header,\"\\n\\n\\n\")\n",
    "                                            pass\n",
    "                            if len(used_headers_faa)==len(all_headers_faa)==expected_len :\n",
    "                                outfile_c.write(\"Seems like this one is good to go :\",specie, strain)\n",
    "                            else :\n",
    "                                outfile_c.write(f\"Possible trouble with the following : {specie} {strain},. Len faa file : {len(faa_file)}. And len of the .gen file :,{len(used_headers_faa)}\")\n",
    "                                print(\"Possible trouble with the following :\",specie, strain,\". Len faa file :\",len(faa_file), \", And len of the .gen file :\",len(used_headers_faa))\n",
    "                                for test_header in all_headers_faa :\n",
    "                                    if test_header not in used_headers_faa :\n",
    "                                        print(test_header)\n",
    "\n",
    "                            \n",
    "# Fixing the last ones :\n",
    "\n",
    "wrong_genes={}\n",
    "for file in os.listdir(f\"{path_data}/Genes\"):\n",
    "    strain=file.split(\".gen\")[0]\n",
    "    cont_genes=[]\n",
    "    wrong_genes[strain]=cont_genes\n",
    "    faa_file=open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\")[1:]\n",
    "    ffn_file=open(f\"{path_data}/Genes/{strain}.gen\").read().split(\">\")[1:]\n",
    "    for index_gen, seq_gen in enumerate(ffn_file):\n",
    "        header_seq=seq_gen.split(\"\\n\")[0]\n",
    "        for index_prt, seq_prot in enumerate(faa_file):\n",
    "            header_prt=seq_prot.split(\"\\n\")[0]\n",
    "            if header_prt == header_seq :\n",
    "                fasta_ffn=\"\".join(seq_gen.split(\"\\n\")[1:])\n",
    "                fasta_faa=\"\".join(seq_prot.split(\"\\n\")[1:])\n",
    "                translation=str(Seq(fasta_ffn).translate())[0:-1].strip()\n",
    "                if translation == fasta_faa :\n",
    "                    pass\n",
    "                else :\n",
    "                    wrong_genes[strain].append(header_seq)\n",
    "                    print(\"That one is wrong :\", strain , header_seq ,\"\\n\", fasta_faa , \"\\n\",translation)\n",
    "                    continue\n",
    "    print(\"Strain :\",strain, \"finshed\")\n",
    "                        \n",
    "wrong_genes={}\n",
    "for file in os.listdir(f\"{path_data}/Genes\"):\n",
    "    strain=file.split(\".gen\")[0]\n",
    "    cont_genes=[]\n",
    "    wrong_genes[strain]=cont_genes\n",
    "    faa_file=open(f\"{path_data}/Proteins/{strain}.prt\").read().split(\">\")[1:]\n",
    "    ffn_file=open(f\"{path_data}/Genes/{strain}.gen\").read().split(\">\")[1:]\n",
    "    for index_gen, seq_gen in enumerate(ffn_file):\n",
    "        header_seq=seq_gen.split(\"\\n\")[0]\n",
    "        for index_prt, seq_prot in enumerate(faa_file):\n",
    "            header_prt=seq_prot.split(\"\\n\")[0]\n",
    "            if header_prt == header_seq :\n",
    "                fasta_ffn=\"\".join(seq_gen.split(\"\\n\")[1:])\n",
    "                fasta_faa=\"\".join(seq_prot.split(\"\\n\")[1:])\n",
    "                translation=str(Seq(fasta_ffn).translate())[0:-1].strip()\n",
    "                if len(translation) == len(fasta_faa) :\n",
    "                    print(\"Might work\")\n",
    "                    pass\n",
    "                else :\n",
    "                    wrong_genes[strain].append(header_seq)\n",
    "                    print(\"That one is wrong :\", strain , header_seq ,\"\\n\",fasta_faa ,\"\\n\",translation)\n",
    "                    continue\n",
    "    print(\"Strain :\",strain, \"finshed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 7. Infer the phylogenetic tree from the previous alignment\n",
    "The tree has been computed directly with iqtree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Local tree : \n",
    "path_local=\"/home/conchae/prediction_depolymerase_tropism/iqtree_local\"\n",
    "\n",
    "\n",
    "iqtree \n",
    "-s /home/conchae/prediction_depolymerase_tropism/iqtree_local/Klebsiella_genomes.nucl.grp.aln -m MFP  --prefix Klensiella_genomes_MFP -alrt 1000  -T AUTO  -B -v\n",
    "\n",
    "iqtree \n",
    "-s /home/conchae/prediction_depolymerase_tropism/iqtree_local/Klebsiella_genomes.nucl.grp.aln\n",
    "-m GTR+F+I\n",
    "--prefix Klensiella_genomes_MFP\n",
    "-T AUTO \n",
    "-B 1000 \n",
    "-alrt 1000\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=MFP_local_tree\n",
    "#SBATCH --partition=long \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=60\n",
    "#SBATCH --mem=200gb \n",
    "#SBATCH --time=10-00:00:00 \n",
    "#SBATCH --output=MFP_local_tree%j.log \n",
    "\n",
    "module restore la_base\n",
    "source /storage/apps/ANACONDA/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate Treebuilding\n",
    "\n",
    "iqtree -s /home/conchae/prediction_depolymerase_tropism/iqtree_local/Klebsiella_genomes.nucl.grp.aln -m MFP  --prefix Klensiella_genomes_MFP -T AUTO  -B 1000 \n",
    "\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=Fixed_tree\n",
    "#SBATCH --partition=long \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=60\n",
    "#SBATCH --mem=200gb \n",
    "#SBATCH --time=10-00:00:00 \n",
    "#SBATCH --output=Fixed_Tree%j.log \n",
    "\n",
    "module restore la_base\n",
    "source /storage/apps/ANACONDA/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate Treebuilding\n",
    "\n",
    "iqtree  \n",
    "-s /home/conchae/prediction_depolymerase_tropism/iqtree_local/Klebsiella_genomes.nucl.grp.aln  -m GTR+F+I  --prefix Klensiella_genomes_fixed  -B 1000  -alrt 1000 -t BIONJ /home/conchae/prediction_depolymerase_tropism/iqtree_local/script_files/Klensiella_genomes_fixed.bionj -nt AUTO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Writting tips file for the tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Writting the tips file :\n",
    "\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "import random\n",
    "\n",
    "path_klebsiella=\"/home/conchae/prediction_depolymerase_tropism\"   \n",
    "\n",
    "with open(f\"{path_klebsiella}/pastML/pastML_KL_tips.comma.txt\", \"w\") as outfile :\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie[0]==\"k\" and os.path.isdir(f\"{path_klebsiella}/{specie}\")== True:\n",
    "            for rep in os.listdir(f\"{path_klebsiella}/{specie}/refseq/bacteria\"):\n",
    "                kaptive_out=open(f\"{path_klebsiella}/{specie}/refseq/bacteria/{rep}/{rep}_Kaptive_out.txt\").read()\n",
    "                first_line=kaptive_out.split(\"\\n\")[0].split(\"\\t\")\n",
    "                secon_line=kaptive_out.split(\"\\n\")[1].split(\"\\t\")\n",
    "                kaptive_dic=dict(zip(first_line,secon_line))\n",
    "                if kaptive_dic[\"Match confidence\"] != \"None\" and kaptive_dic[\"Match confidence\"] != \"Low\":\n",
    "                    outfile.write(f\"{rep},{kaptive_dic['Best match locus']}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Check how well supported the tree is"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how wel the tree is supported :\n",
    "\n",
    "\n",
    "from os import system, listdir, chdir, mkdir\n",
    "from os.path import isdir\n",
    "import os\n",
    "\n",
    "path_tree=\"/home/conchae/prediction_depolymerase_tropism/iqtree_local/tree_files\"\n",
    "\n",
    "tree=[value.split(\":\")[0] for value in open(f\"{path_tree}/Klensiella_genomes_fixed.2.1.treefile\").read().split(\")\")]\n",
    "\n",
    "sh_alrt=[float(num.split(\"/\")[0]) for num in tree if num.count(\"/\")>0 and float(num.split(\"/\")[0])!=0.0]\n",
    "bootstrap=[float(num.split(\"/\")[1]) for num in tree if num.count(\"/\")>0 and float(num.split(\"/\")[0])!=0.0]\n",
    "\n",
    "def Average(lst):\n",
    "    return sum(lst) / len(lst)\n",
    "\n",
    "df = pd.Dataframe(bootstrap_values)._convert(numeric=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
