{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run Vcontact2 with the prophage genomes\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate the input files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step one : generating the fasta file with the phages proteins and the gene_to_genome file\n",
    "\n",
    "from os import listdir, mkdir, rename, system\n",
    "import os\n",
    "\n",
    "\n",
    "path_phageboost=\"/home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_decipher/phageboost/phageboost_70\"\n",
    "path_phylogeny=\"/home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_phylogeny\"\n",
    "\n",
    "\n",
    "\n",
    "with open(f\"{path_phylogeny}/phageboost_70_Vcontact2.fasta\",\"w\") as outfile :\n",
    "    with open(f\"{path_phylogeny}/phageboost_70_gene_to_genome.csv\",\"w\") as outfile2:\n",
    "        outfile2.write(\"protein_id,contig_id,keywords\\n\")\n",
    "        for rep in os.listdir(path_phageboost) :\n",
    "            for strain in os.listdir(f\"{path_phageboost}/{rep}\") :\n",
    "                for file in os.listdir(f\"{path_phageboost}/{rep}/{strain}\") :\n",
    "                    if file[-6:]==\".fasta\":\n",
    "                        prophage_name=file.split(\".fasta\")[0]\n",
    "                        prot_file=open(f\"{path_phageboost}/{rep}/{strain}/{prophage_name.split('__')[-1]}.multi.faa\").read()\n",
    "                        outfile.write(f\"{prot_file}\\n\")\n",
    "                        prot_file_id=[seq.split(\"\\n\")[0] for seq in prot_file.split(\">\")]\n",
    "                        for index_seq, seq_id in enumerate(prot_file_id):\n",
    "                            if seq_id:\n",
    "                                outfile2.write(f\"{seq_id},{prophage_name},None_provided\\n\")\n",
    "                \n",
    "\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=writting_Vcon\n",
    "#SBATCH --partition=short \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=10\n",
    "#SBATCH --mem=75gb \n",
    "#SBATCH --time=1-00:00:00 \n",
    "#SBATCH --output=writting_Vcon%j.log \n",
    "\n",
    "module restore la_base\n",
    "                \n",
    "                \n",
    "python3 /home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_phylogeny/script_file/py_files/generate_vcontact2_files.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the previously created input files with the Millard data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import listdir, mkdir, rename, system\n",
    "import os\n",
    "\n",
    "\n",
    "path_phageboost=\"/home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_decipher/phageboost/phageboost_70\"\n",
    "path_phylogeny=\"/home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_phylogeny\"\n",
    "path_77=\"/home/conchae/77_strains_phage_project\"\n",
    "path_millard = \"/home/conchae/databases/Millard_sept_2022\"\n",
    "\n",
    "gene2genome_m = \"\\n\".join(open(f\"{path_millard}/1Sep2022_vConTACT2_gene_to_genome.csv\").read().split(\"\\n\")[1:])\n",
    "gene_seq_m = open(f\"{path_millard}/1Sep2022_vConTACT2_proteins.faa\").read()\n",
    "\n",
    "PPT_gene2genome = open(f\"{path_phylogeny}/phageboost_70_gene_to_genome.csv\").read()\n",
    "PPT_gene_seq = open(f\"{path_phylogeny}/phageboost_70_Vcontact2.fasta\").read()\n",
    "\n",
    "with open(f\"{path_phylogeny}/phageboost_70_Vcontact2.Millard.fasta\",\"w\") as outfile :\n",
    "    with open(f\"{path_phylogeny}/phageboost_70_gene_to_genome.Millard.csv\",\"w\") as outfile2:\n",
    "        outfile.write(f\"{PPT_gene_seq}\\n{gene_seq_m}\")\n",
    "        outfile2.write(f\"{PPT_gene2genome}\\n{gene2genome_m}\")\n",
    "        \n",
    "        \n",
    "        \n",
    "# ************************************************************************************************************************************************       \n",
    "#!/bin/bash\n",
    "#BATCH --job-name=writting_Vcon\n",
    "#SBATCH --partition=short \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=10\n",
    "#SBATCH --mem=35gb \n",
    "#SBATCH --time=1-00:00:00 \n",
    "#SBATCH --output=writting_Vcon%j.log \n",
    "\n",
    "module restore la_base\n",
    "                \n",
    "python3 /home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_phylogeny/script_file/py_files/generate_vcontact2_files.Millard.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step two : The actual command line :\n",
    "\n",
    "#!/bin/bash\n",
    "#BATCH --job-name=vcontact2_phageboost\n",
    "#SBATCH --partition=long \n",
    "#SBATCH --ntasks=1 \n",
    "#SBATCH --cpus-per-task=60\n",
    "#SBATCH --mem=250gb \n",
    "#SBATCH --time=2-00:00:00 \n",
    "#SBATCH --output=vcontact2_phageboost%j.log \n",
    "\n",
    "module restore la_base\n",
    "source /storage/apps/ANACONDA/anaconda3/etc/profile.d/conda.sh\n",
    "conda activate vcontact2\n",
    "\n",
    "\n",
    "vcontact  --raw-proteins /home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_phylogeny/phageboost_70_Vcontact2.Millard.fasta --rel-mode 'Diamond' --proteins-fp /home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_phylogeny/phageboost_70_gene_to_genome.Millard.csv --db 'None'  --pcs-mode MCL --vcs-mode ClusterONE --c1-bin /home/conchae/software/cluster_one-1.0.jar --output-dir  /home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_phylogeny/Vcontact2_PPTmillard_output_06102022 -t 60\n",
    "\n",
    "\n",
    "\n",
    "--blast-fp /home/conchae/prediction_depolymerase_tropism/prophage_prediction/prophage_phylogeny/Vcontact2_PPTmillard_output_06102022/phageboost_70_Vcontact2.Millard.self-diamond.tab "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import listdir, mkdir, rename, system\n",
    "import os\n",
    "\n",
    "path_77=\"/home/conchae/77_strains_phage_project\"\n",
    "path_77_seq=\"/home/conchae/77_strains_phage_project/phages_sequences_corrected\"\n",
    "\n",
    "vcontact2_results=open(path_77+\"/vConTACT2_77_phages_V5/genome_by_genome_overview.csv\").read().split(\"\\n\")\n",
    "reference={}\n",
    "our_phages=[]\n",
    "\n",
    "for phage in os.listdir(path_77_seq):\n",
    "    our_phages.append(phage)\n",
    "    \n",
    "for n in range (1,len(vcontact2_results)-1):\n",
    "    phage=vcontact2_results[n].split(\",\")[1]\n",
    "    if phage not in our_phages:\n",
    "        V_C=vcontact2_results[n].split(\",\")[5]\n",
    "        order=vcontact2_results[n].split(\",\")[2]\n",
    "        family=vcontact2_results[n].split(\",\")[3]\n",
    "        genus=vcontact2_results[n].split(\",\")[4]\n",
    "        score=vcontact2_results[n].split(\",\")[-1]\n",
    "        if V_C !=\"\" and V_C not in reference and float(score) > 0.9 :\n",
    "            a={\"Order\":order}\n",
    "            b={\"Family\":family}\n",
    "            c={\"Genus\":genus}\n",
    "            a.update(b)\n",
    "            a.update(c)\n",
    "            reference[V_C]=a\n",
    "\n",
    "with open(path_77+\"/Vcontact_77_results.tsv\",\"w\") as outfile:\n",
    "    for n in range(1,len(vcontact2_results)-1):\n",
    "        phage=vcontact2_results[n].split(\",\")[1]\n",
    "        if phage in our_phages:\n",
    "            V_C=vcontact2_results[n].split(\",\")[5]\n",
    "            score=vcontact2_results[n].split(\",\")[-1]\n",
    "            try :\n",
    "                if float(score) > 0.9 and V_C in reference:\n",
    "                    outfile.write(phage+\"\\t\"+reference[V_C][\"Order\"]+\"\\t\"+reference[V_C][\"Family\"]+\"\\t\"+reference[V_C][\"Genus\"]+\"\\t\"+V_C+\"\\t\"+score+\"\\n\")\n",
    "                elif float(score) > 0.7 and V_C in reference :\n",
    "                    outfile.write(phage+\"\\t\"+reference[V_C][\"Order\"]+\"\\t\"+reference[V_C][\"Family\"]+\"\\t\"+V_C+\"\\t\"+score+\"\\n\")\n",
    "                elif V_C ==\"67_0\":\n",
    "                    outfile.write(phage+\"\\t Caudovirales \\t Siphoviridae \\n\")\n",
    "                elif V_C not in reference :\n",
    "                    print(phage, V_C)\n",
    "                    outfile.write(vcontact2_results[n]+\"\\n\")\n",
    "                else:\n",
    "                    print(phage)\n",
    "            except ValueError:\n",
    "                outfile.write(vcontact2_results[n]+\"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
